### ISSUE TITLE
Qwen3.5-4B (tied embeddings + Qwen3_5 arch): loads, reports tying success, then emits uniform logits (PPL == vocab_size)

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1282]
`Qwen3.5-4B` (BF16, straight from `Qwen/Qwen3.5-4B`) loads without an error and then emits `<|im_start|>` forever. Teacher-forced perplexity is **248320.0000**, which is exactly `vocab_size` — `mean_nll = 12.4225 = ln(248320)`. That number is the signature of perfectly uniform logits: the output projection produces nothing.

Found while trying to build a BF16 hybrid checkpoint to validate the #1273 quantizer fix against.

## Repro

```
imp-cli --model Qwen3.5-4B --prompt "The capital of France is" --max-tokens 12 --temperature 0
  -> [tok=248045 '<|im_start|>'] x12
imp-cli --model Qwen3.5-4B --perplexity tools/analysis/ppl_corpus_45k.txt
  -> perplexity: 248320.0000   (vocab_size = 248320)
```

The BF16 source does this, so it is not a quantization effect — an NVFP4 export built from it behaves the same way.

## What it is not

The load path looks healthy in the log, which is what makes this quiet:

```
WARN  weight_map.cpp:1400: output projection (out_proj) was not found (may be tied to embed_tokens)
INFO  safetensors_loader.cpp:995: Tied output projection to token embedding
INFO  weight_upload.cu:949: Output projection shares GPU data with token embedding (weight tying)
INFO  weight_map.cpp:1320: WeightMap (qwen35): assigned 426 tensors, skipped 312, layers=32, experts=0
```

So the tying fallback *fires* and reports success. Tied embeddings alone are also not the trigger — **Qwen3-0.6B** (`tie_word_embeddings: true`) generates correctly on the same build.

## Narrowed

| model | architecture | `tie_word_embeddings` | vision | generates |
|---|---|---|---|---|
| Qwen3-0.6B | `Qwen3ForCausalLM` | **true** | no | ✅ |
| Qwen3.6-27B-Text-NVFP4 | `Qwen3_5ForConditionalGeneration` | false (real `lm_head`) | no | ✅ |
| Qwen3.6-35B-A3B-NVFP4 | `Qwen3_5MoeForConditionalGeneration` | false | yes | ✅ |
| **Qwen3.5-4B** | `Qwen3_5ForConditionalGeneration` | **true** | yes | ❌ |

The failing case is the only one that combines a `Qwen3_5*` architecture with tied embeddings. Under that architecture the tensor is named `model.language_model.embed_tokens.weight` (note the `language_model.` infix) rather than `model.embed_tokens.weight`, so the same fallback that works for `Qwen3ForCausalLM` may be tying a pointer that is not the tensor the LM head path actually reads.

One possibly related line from the same load:

```
INFO  pre_dequant_phase4_tensor_registry.cu:348: Phase-4 plan/actual MISMATCH: TOK_EMBED plan-tier=2 actual-tier=5
```

## Why it matters beyond one checkpoint

`PPL == vocab_size` is worth knowing as a diagnostic on its own: it means uniform logits, i.e. the head is dead, and it is distinguishable at a glance from "quantization made this worse". It is also **silent** — no error, no refusal, and a plausible-looking load summary.

This is the third instance of the class where a model loads, reports success, and serves garbage (cf. the vision tower loading 247/316 and answering anyway).

**Comment by kekzl** [comment]
**Not the lm_head — the forward produces NaN. Localised, and a separate real bug found on the way that does not explain it.**

## Where it breaks

Per-layer prefill dumps, `"Hello world"`, 12 tokens:

| layer | state |
|---|---|
| L00–L07 | healthy, RMS grows normally, 0 NaN |
| **L08** `A_pre_attn` | **0 NaN** — the block's input is clean |
| **L08** `B_post_attn` | **NaN** |
| L09–L10 | same partial NaN, carried |
| L11+ | entirely non-finite |

So the NaN is created **inside the block of layer 8**, which is a `linear_attention` (GDN) layer. My earlier guess — tied embeddings / the LM head — is **wrong**: the tying fallback fires correctly, and `PPL == vocab_size` is a downstream consequence of NaN logits, not the cause.

## The sharp part: a fixed token boundary

NaN starts at **token position 6**, and that boundary does not move with prompt length:

| prompt tokens | NaN tokens | first bad |
|---|---|---|
| 11 | 5 | 6 |
| 12 | 6 | 6 |
| 23 | 17 | 6 |

Tokens 0–5 are clean, everything from 6 on is fully NaN (2560/2560 features). A numerical overflow would not produce a boundary this sharp, and it is not proportional — so this looks like an addressing/capacity fault, not saturation.

## Ruled out

- **Missing weights.** 441 non-vision tensors, 426 assigned; the 15 unassigned are exactly the MTP tensors.
- **Weights outside FP16 range.** Every layer-8 tensor has absmax ≤ 7.22 (FP16 max is 65504).
- **A regression from today's PRs.** 0.21.0 (prebuilt, before #1278–#1281) shows the identical failure.
- **A shard boundary at layer 8.** Layers 6–9 all straddle both shards.
- **Tied embeddings as such.** Qwen3-0.6B (`tie_word_embeddings: true`) generates correctly.

## Found and fixed on the way — but it is NOT this bug

`has_gdn_` was assigned *after* `compute_shared_sizes()` ran, so the shared arena was sized with `proj_elem_size = 2` (FP16) while `configure_ssm_workspace()` later carved with `4` (FP32), the flag by then being set. On this model:

```
before:  Shared workspace: 244.50 MiB = max(attn=164.0, ffn=236.0, moe=0.0, ssm=244.5)
after:   Shared workspace: 437.50 MiB = max(attn=164.0, ffn=236.0, moe=0.0, ssm=437.5)
```

**193 MiB under-reserved**, with pointers handed out past the end. A GDN+MoE model hides this because the MoE phase dominates the arena; a GDN+dense model like this one has the SSM phase as the maximum, so it is exposed. Same class as the `has_gdn_`-read-before-assignment already documented for the max_tokens cap (AUDIT B18).

Fixing it changes the reservation as shown above and **does not change the NaN at all** — same layers, same token boundary. Worth shipping on its own; recorded here so nobody re-derives it as the cause.

## Next

Instrumentation inside the GDN block (post-conv1d, post-scan, pre-out_proj) to find which step first sees position 6. The token-6 boundary is the strongest lead: something is sized or strided for 6 rows where 12 arrive.

**Comment by kekzl** [comment]
**Root cause: the GDN delta-rule scan diverges exponentially and overflows FP16 at token 6.**

The existing GDN dumps already carry every intermediate, so no new instrumentation was needed. Layer 8, `"Hello world"`, per-token absmax:

| stage | t0 | t1 | t2 | t3 | t4 | t5 | t6 |
|---|---|---|---|---|---|---|---|
| **`gdn_y_post_scan`** | 0.04 | 0.06 | 0.40 | 2.51 | **110.06** | **31680.00** | **inf** |
| `gdn_y_post_norm` | 0.92 | 1.35 | 1.08 | 2.64 | 2.16 | 2.44 | 1.75 (9 NaN) |

Growth per token: 1.5x, 6.7x, 6.3x, 44x, 288x — accelerating, i.e. a state multiplying itself rather than decaying. It crosses 65504 between t5 and t6 and becomes `inf`; `rmsnorm(inf)` then yields NaN, which is why `gdn_y_post_norm` is where NaN first appear (9, then 109, then 125).

Everything upstream is clean: `gdn_ssm_in_out`, `gdn_conv_f32`, `gdn_alpha`, `gdn_beta` and the scan's own output all have **0 NaN**. The scan output is finite until it simply gets too large.

## Correcting my earlier reading

I described a "hard boundary at token 6, independent of prompt length" and inferred an addressing or capacity fault. That was wrong. The boundary is sharp *downstream* only because the `ssm_out` GEMM turns one NaN into a full row — at the point of creation it is gradual (9 → 109 → 125 NaN). The real signal is the absmax curve above, which I had not looked at per token.

## Also not the cause (fixed separately)

`has_gdn_` was read before assignment when sizing the shared arena, under-reserving it by 193 MiB on this model (244.5 → 437.5 MiB). Real bug, shipped as #1283 — it does not change this NaN at all.

## Where to look next

A decay factor that should be < 1. In gated delta-rule the state decays by `exp(a * dt)` with `a = -exp(A_log)`, i.e. `a` must be negative. A sign or activation missing on that path would produce exactly this: a state that grows instead of decays, at a rate that compounds.

Worth checking why it survives on other GDN checkpoints — `Qwen3.5-4B-mxfp4.gguf` (same architecture, quantized) generates correctly, so the scan itself works there. The difference to look at first is how `A_log`/`dt_bias` reach the kernel on the BF16 SafeTensors path versus the GGUF path.

**Comment by kekzl** [comment]
**Root cause confirmed: `A_log` is used raw on the SafeTensors path. It needs `-exp()`.**

Comparing the two load paths on the *same model* (Qwen3.5-4B), layer 8:

| source | tensor | range | all negative |
|---|---|---|---|
| **GGUF** (mxfp4, works) | `blk.8.ssm_a` | **-7.7437 … -0.0681** | **yes** |
| **SafeTensors** (BF16, broken) | `linear_attn.A_log` | -2.6875 … **+2.0469** | no |
| computed `-exp(A_log)` | — | **-7.7437 … -0.0681** | yes |

`-exp(A_log)` reproduces the GGUF values **to four decimals**. So llama.cpp's converter applies the transform at export time, and imp consumes `ssm_a` already transformed. The SafeTensors loader maps `A_log -> ssm_a` verbatim (`weight_map.cpp:295`, `:1100`) without it.

In gated delta-rule the state decays by `exp(a * dt)`, so `a` must be negative. Feeding raw `A_log` hands the scan positive rates up to +2.05, turning decay into growth. That is exactly the measured curve — per-token absmax of the scan output: 0.04, 0.06, 0.40, 2.51, 110.06, 31680.00, inf. It crosses FP16 between t5 and t6, and `rmsnorm(inf)` makes the NaN.

Every other scan input is identical across the two paths, which is what isolates it:

| stage | BF16 | GGUF |
|---|---|---|
| `gdn_ssm_in_out` | -35.13 … 16.06 | -36.72 … 15.95 |
| `gdn_conv_f32` | **-0.2785** … 10.83 | **-0.2785** … 11.29 |
| `gdn_alpha` | -2.06 … 6.67 | -2.21 … 6.82 |
| `gdn_y_post_scan` | **-inf … inf** | **-0.041 … 0.032** |

## Fix

Apply `a = -exp(A_log)` when loading `A_log` from SafeTensors, so both paths hand the kernel the same quantity. The GGUF path must stay as-is — its values are already transformed, and applying it twice would give `-exp(-exp(x))`, i.e. a decay near -1 for everything.

That asymmetry is worth a comment at the mapping site: the two formats store *different quantities under the same name*, and nothing about `ssm_a` says which one you have.

**Comment by kekzl** [comment]
Fixed. The tying was never the problem — the narrowing table in this issue was right that the failing case is the only `Qwen3_5*` + tied-embeddings combination, but that correlation pointed at the wrong tensor.

The cause was **`A_log`** (#1284, merged in the #1285 squash): imp's GDN scan expects the post-`-exp()` value that llama.cpp's converter bakes in, and the upload path decided "is this HF?" from the **dtype** — treating F32 as "GGUF, already transformed". Qwen3.5-4B stores `A_log` as F32. Its raw values reached the scan as a positive decay rate, so the recurrent state grew instead of decaying:

```
per-token absmax: 0.04, 0.06, 0.40, 2.51, 110.06, 31680.00, inf
```

It crosses FP16 between token 5 and 6, `rmsnorm(inf)` yields NaN, and every layer past 11 is non-finite. Uniform logits — hence `PPL == vocab_size` — are what a NaN-poisoned final hidden state produces, and the tying log lines were healthy because the tying *was* healthy.

Decided from the values instead, which is decidable: `-exp(x)` is strictly negative for every real `x`, so any value ≥ 0 can only be a raw HF `A_log`.

Verified on current `main` (`120bc0d7`), the repro from this issue verbatim:

```
imp-cli --model Qwen3.5-4B-BF16 --prompt "The capital of France is" --max-tokens 12 --temperature 0
  -> "Thinking Process:\n\n1. **Analyze the ..."   (token ids 90700, 8340, 25, 271, 16, ... — not 248045 twelve times)
```

Perplexity over `ppl_corpus_45k.txt`: **248320.0000 → 9.2624**. The second half of that improvement is #1289, which fixed the final RMSNorm missing Qwen3.5/3.6's `gamma = 1 + W` offset — a separate defect this checkpoint also had, and the one that turned out to be #1273.

Worth recording for the next reader: **two independent defects sat on the same checkpoint**, and the first one's signature (uniform logits) completely masked the second (a 41% perplexity gap against an HF reference). Fixing one made the other measurable for the first time.
